azkaban-web-server/src/main/java/cloudflow/servlets/ProjectServlet.java

```diff
@@ -0,0 +1,100 @@
+  private static final UriTemplate GET_PROJECT_URI_TEMPLATE = new UriTemplate(
+      String.format("/projects/{%s}", PROJECT_ID_KEY));
+
+  private ProjectService projectService;
+  private ObjectMapper objectMapper;
+
+  private static final Logger log = LoggerFactory.getLogger(ProjectServlet.class);
+
+  @Override
+  public void init(final ServletConfig config) throws ServletException {
+    super.init(config);
+    final AzkabanWebServer server = (AzkabanWebServer) getApplication();
+    this.projectService = server.projectService();
+    this.objectMapper = server.objectMapper();
+  }
+
+  @Override
+  protected void handleGet(HttpServletRequest req, HttpServletResponse resp, Session session)
+      throws IOException, ServletException {
+    Map<String, String> templateVariableToValue = new HashMap<>();
+    if (GET_ALL_PROJECT_URI.equals(req.getRequestURI())) {
+      /* Get all records */
+      String response = objectMapper.writeValueAsString(projectService.getAllProjects(session.getUser()));
+      this.writeResponse(resp, response);
+      return;
+    } else if (GET_PROJECT_URI_TEMPLATE.match(req.getRequestURI(), templateVariableToValue)) {
+      /* Get specific record */
+      try {
+        String projectId = templateVariableToValue.get(PROJECT_ID_KEY);
+
+        /* Validate projectId is an Integer. */
+        try {
+          Integer.parseInt(projectId);
+        } catch (NumberFormatException e) {
+          log.error("Invalid project id: ", projectId);
+          sendErrorResponse(resp, HttpServletResponse.SC_BAD_REQUEST, "Invalid project id");
+          return;
+        }
+
+        String response = objectMapper.writeValueAsString(projectService.getProject(projectId));
+        this.writeResponse(resp, response);
+      } catch (Exception e) {
+        log.error("Exception while fetching project: " + e);
+        sendErrorResponse(resp, HttpServletResponse.SC_NOT_FOUND, "Project id not found");
+        return;
+      }
+    } else {
+      /* Unsupported route, return an error */
+      log.error("Invalid route for projects endpoint: " + req.getRequestURI());
+      sendErrorResponse(resp, HttpServletResponse.SC_NOT_IMPLEMENTED, "Unsupported projects API "
+          + "endpoint");
+      return;
+    }
+  }
+
+  @Override
+  protected void handlePost(HttpServletRequest req, HttpServletResponse resp, Session session)
+      throws ServletException, IOException {
+    try {
+      String body = HttpRequestUtils.getBody(req);
+      Project project = objectMapper.readValue(body, Project.class);
+
+      String projectId = projectService.createProject(project, session.getUser());
+      sendResponse(resp, HttpServletResponse.SC_CREATED, projectId);
+      return;
+    } catch(Exception e) {
+      log.error("Error while handling POST request for projects. Got Exception: " + e);
+      sendErrorResponse(resp, HttpServletResponse.SC_BAD_REQUEST, "BAD request to create project");
+      return;
+    }
+  }
```

Comment:
Can you catch validation errors separately so that unexpected errors (500) thrown by the service or the layers below are not reported as bad request (400)?

Reply:
Will improve exception handling in follow-up change since other changes are waiting on this.